### PR TITLE
feat(controller): Add validation for ContainerSet

### PR DESCRIPTION
**Pull request by dinever**
The `ContainerSet` currently lacks validation.  This PR adds it.

- [x] Validate container set must have at least one container:
    ```
    templates.main.containerSet.containers must have at least one container
    ```
- [x] Validate container name uniqueness:
    ```
    templates.main.containerSet.containers[1].name 'a' is not unique
    ```
- [x] Validate dependency name exits:
    ```
    templates.main.containerSet.containers.b dependency 'c' not defined
    ```
- [x] Validate no dependency cycles:
    ```
    templates.main.containerSet.containers dependency cycle detected: b->a->b
    ```

Added full test coverage.

### CODE REVIEW

File: workflow/validate/validate.go [hunk 1]
```diff
@@ -1017,6 +1049,28 @@ func validateOutputParameter(paramRef string, param *wfv1.Parameter) error {
...
+		typeField := s.Type().Field(i)
+		if typeField.Anonymous {
+			name, err := getNameFieldValue(s.Field(i))
+			if err == nil {
+				return name, nil
+			}
+		}
+	}
+	return "", errors.InternalError("No 'Name' field in struct")
+}
```
Comment (dinever) [inline workflow/validate/validate.go:1072]: The original `getNameFieldValue` doesn't support `Name` field being embedded in another struct. So I updated it in order to use it for `ContainerNode` struct. For example:

```go
type Container struct {
	Name string
}

type ContainerNode struct {
	Container
}
```

Moved the func out so that I can write unit tests for it.

File: workflow/validate/validate.go [hunk 2]
```diff
@@ -1290,21 +1334,51 @@ func validateDAGTargets(tmpl *wfv1.Template, nameToTask map[string]wfv1.DAGTask)
...
+	}
+	err := verifyNoCycles(depGraph)
+	if err != nil {
+		return errors.Errorf(errors.CodeBadRequest, "templates.%s.containerSet.containers %s", tmpl.Name, err.Error())
+	}
+	return nil
+}
+
+// verifyNoCycles verifies that a dependency graph has no cycles by Depth-First Search
+// depGraph is an adjacency list, where key is a node name and value is a list of its dependencies' names
```
Comment (dinever) [inline workflow/validate/validate.go:1364]: Had to refactor this func a little bit so that it can be shared between `DAG` and `ContainerSet`.

File: workflow/validate/validate.go [hunk 3]
```diff
@@ -1017,6 +1049,28 @@ func validateOutputParameter(paramRef string, param *wfv1.Parameter) error {
 	return nil
 }
 
+// getNameFieldValue gets the value of the `Name` field from the struct
+// if not found from the struct, it checks embedded field instead
+func getNameFieldValue(s reflect.Value) (string, error) {
```
Comment (alexec) [inline workflow/validate/validate.go:1054]: Reflection can make bugs hard to find and fix. You cannot perform refracting when using reflection for example. So you write more tests. Is there any way we can avoid it?

File: workflow/validate/validate.go (hunk 2 above)
Comment (alexec) [reply workflow/validate/validate.go:1364]: Nice

**Review by alexec — COMMENTED** [review]
* Good stuff. I omitted these as they were not MVP, but they’re useful.
* I feel like validation should be intrinsic, e.g. in a function named ContainerSet.Validate(). The way we do it is far away from the code.
* Another test we might want to add sometime: if there are inputs or outputs you must man a container named “main”.

File: workflow/validate/validate.go (hunk 3 above)
Comment (dinever) [reply workflow/validate/validate.go:1054]: Thanks! Totally agree. I think to avoid reflect completely, the most straight-forward way is to just make the validation method accept `[]string` of names instead:

```diff
--- a/workflow/validate/validate.go
+++ b/workflow/validate/validate.go
@@ -1022,7 +1022,7 @@ func validateOutputParameter(paramRef string, param *wfv1.Parameter) error {
 // * unique
 // * non-empty
 // * matches matches our regex requirements
-func validateWorkflowFieldNames(slice interface{}) error {
+func validateWorkflowFieldNames(names []string, isParamOrArtifact bool) error {
```

Note that we need `isParamOrArtifact` because parameters and artifact has different name validation regex than other fields.

When we use it, we create a slice of names instead:

```go
names := []string{}
for _, t := range wf.Spec.Templates {
	names = append(names, t.Name)
}
validateWorkflowFieldNames(names, false)
```

What do we think?

**Comment by dinever** [comment]
Thanks @alexec !

> I feel like validation should be intrinsic, e.g. in a function named ContainerSet.Validate(). The way we do it is far away from the code.

Sounds good! I can move the validation into `pkg/apis/workflow/v1alpha1/container_set_template_types.go`. However this will introduce some inconsistency. I can work on refactoring other validation as well in future PRs until they are fully migrated.

File: workflow/validate/validate.go (hunk 3 above)
Comment (alexec) [reply workflow/validate/validate.go:1054]: I agree

**Comment by alexec** [comment]
don't worry about refactoring existing code - the existing validation is a bit buggy - so it is best we try and make sure new validations are really tight

**Comment by dinever** [comment]
Thanks for the review @alexec ! I've pushed the latest changes

Changes:
1. Moved validation for `ContainerSet` into `ContainerSet.Validate`

    However, since we also need to validate `main` container when there is input/output, input/output are not not within `ContainerSet`, so I have to pass in the `Template` as well for validation:

    ```go
    func (in *ContainerSetTemplate) Validate(tmpl *Template) error 
    ```

2. Copied & modified some utility functions into `pkg/apis/workflow/v1alpha1/validation_utils.go`, e.g. `validateNoCycles` and `validateWorkflowFieldNames`. I feel it probably doesn't make sense for `v1alpha1` to expose a utility function for the controller to call, so I kept the original util funcs in `validate.go` as is. I can work on migrating other validations in later PRs.

Thanks!

File: pkg/apis/workflow/v1alpha1/container_set_template_types.go [hunk 4]
```diff
@@ -42,6 +46,62 @@ func (in *ContainerSetTemplate) HasSequencedContainers() bool {
...
+	names := make([]string, 0)
+	for _, ctr := range in.Containers {
+		names = append(names, ctr.Name)
+	}
+	err := validateWorkflowFieldNames(names, false)
+	if err != nil {
+		return fmt.Errorf("containers%s", err.Error())
+	}
+
+	if len(tmpl.Inputs.Artifacts) > 0 || len(tmpl.Outputs.Parameters) > 0 || len(tmpl.Outputs.Artifacts) > 0 {
```
Comment (dinever) [inline pkg/apis/workflow/v1alpha1/container_set_template_types.go:64]: AFAIK it's okay to have `inputs.parameters` without the main container?

For example, this works:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: pod-
  labels:
    workflows.argoproj.io/container-runtime-executor: emissary
spec:
  entrypoint: main
  templates:
  - name: main
    steps:
      - - name: main
          template: print-message
          arguments:
            parameters:
              - name: message
                value: "hello world"

  - name: print-message
    inputs:
      parameters:
        - name: message

    containerSet:
      containers:
        - name: a
          image: argoproj/argosay:v2
          command: ["cowsay"]
          args: ["{{inputs.parameters.message}}"]
```

File: pkg/apis/workflow/v1alpha1/container_set_template_types.go [hunk 5]
```diff
@@ -42,6 +46,62 @@ func (in *ContainerSetTemplate) HasSequencedContainers() bool {
 	return false
 }
 
+// Validate checks if the ContainerSetTemplate is valid
+func (in *ContainerSetTemplate) Validate(tmpl *Template) error {
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/container_set_template_types.go:50]: this seems upside down to me - this should not take any parameters - any validated of the template should be done in `Template.Validate()`

File: pkg/apis/workflow/v1alpha1/container_set_template_types_test.go [hunk 6]
```diff
@@ -5,8 +5,18 @@ import (
 
 	"github.com/stretchr/testify/assert"
 	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/yaml"
 )
 
+func validateContainerSetTemplate(yamlStr string) error {
+	var tmpl Template
+	err := yaml.Unmarshal([]byte(yamlStr), &tmpl)
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/container_set_template_types_test.go:13]: I keep needing a `MustUnmarshal(interface{})` function - let me create a PR to add it

File: pkg/apis/workflow/v1alpha1/container_set_template_types.go (hunk 5 above)
Comment (dinever) [reply pkg/apis/workflow/v1alpha1/container_set_template_types.go]: Thanks for the review @alexec!

I just pushed the latest commit to move the validation out.

File: pkg/apis/workflow/v1alpha1/container_set_template_types_test.go (hunk 6 above)
Comment (dinever) [reply pkg/apis/workflow/v1alpha1/container_set_template_types_test.go]: Great! I can rebase after #5785 to make use of that. Thanks!

File: pkg/apis/workflow/v1alpha1/validation_utils.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
+package v1alpha1
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/validation_utils.go:1]: Is this file still needed?

Comment (dinever) [reply pkg/apis/workflow/v1alpha1/validation_utils.go:1]: Yes, `validateNoCycles` and `validateWorkflowFieldNames` are needed by `ContainerSetTemplate.Validate`. I put them in `validation_utils.go` since other validation may need them later

**Review by alexec — APPROVED** [review]
